src/diagnostics.md

```diff
@@ -0,0 +1,271 @@
+#[forbid(missing_docs)]
+pub mod m3 {
+    // Attempting to toggle warning signals an error here
+    #[allow(missing_docs)]
+    /// Returns 2.
+    pub fn undocumented_too() -> i32 { 2 }
+}
+```
+
+### Tool lint attributes
+
+Tool lints let you use scoped lints, to `allow`, `warn`, `deny` or `forbid` lints of
+certain tools.
+
+Currently `clippy` is the only available lint tool.
+
+Tool lints only get checked when the associated tool is active. If a lint
+attribute, such as `allow`, references a nonexistent tool lint, the compiler
+will not warn about the nonexistent lint until you use the tool.
+
+Otherwise, they work just like regular lint attributes:
+
+```rust
+// set the entire `pedantic` clippy lint group to warn
+#![warn(clippy::pedantic)]
+// silence warnings from the `filter_map` clippy lint
+#![allow(clippy::filter_map)]
+
+fn main() {
+    // ...
+}
+
+// silence the `cmp_nan` clippy lint just for this function
+#[allow(clippy::cmp_nan)]
+fn foo() {
+    // ...
+}
+```
+
+## The `deprecated` attribute
+
+The *`deprecated` attribute* marks an item as deprecated. `rustc` will issue
+warnings on usage of `#[deprecated]` items. `rustdoc` will show item
+deprecation, including the `since` version and `note`, if available.
+
+The `deprecated` attribute has several forms:
+
+- `deprecated` — Issues a generic message.
+- `deprecated = "message"` — Includes the given string in the deprecation
+  message.
+- [_MetaListNameValueStr_] syntax with two optional fields:
+  - `since` — Specifies a version number when the item was deprecated. `rustc`
+    does not currently interpret the string, but external tools like [Clippy]
+    may check the validity of the value.
+  - `note` — Specifies a string that should be included in the deprecation
+    message. This is typically used to provide an explanation about the
+    deprecation and preferred alternatives.
+
+The `deprecated` attribute may be applied to any [item], [trait item], [enum
+variant], [struct field], or [external block item]. It cannot be applied to [trait
+implementation items]. When applied to an item containing other items, such as
+a [module] or [implementation], all child items inherit the deprecation attribute.
+<!-- NOTE: Currently broken for macros, see https://github.com/rust-lang/rust/issues/49912
+Also, it is only rejected for trait impl items (AnnotationKind::Prohibited). In all
+other locations, it is silently ignored. Tuple struct fields are ignored.
+-->
+
+Here is an example:
+
+```rust
+#[deprecated(since = "5.2", note = "foo was rarely used. Users should instead use bar")]
+pub fn foo() {}
+
+pub fn bar() {}
+```
+
+The [RFC][1270-deprecation.md] contains motivations and more details.
+
+[1270-deprecation.md]: https://github.com/rust-lang/rfcs/blob/master/text/1270-deprecation.md
+
+
+## The `must_use` attribute
+
+The *`must_use` attribute* can be used on user-defined composite types
+([`struct`s][struct], [`enum`s][enum], and [`union`s][union]) and [functions].
```

Comment:
I can't quite figure out how to get the must_use lint to trigger for a trait object (it works fine for impl trait).  rust-lang/rust#55506 implies something like `Box<dyn Trait>` would work, but I can't get it to trigger (I think the code that handles it is [here](https://github.com/rust-lang/rust/blame/52e885628e4317aa3f158622435927eb29b812e9/src/librustc_lint/unused.rs#L81-L91)). Can you make an example that triggers for a trait object?

Reply:
That is curious... I cannot produce a warning even with `unsized_locals`.. cc @varkor.

Reply:
Here's a reproducer:
```rust
#![feature(unsized_locals)]

#[must_use] trait Foo {}

impl<T> Foo for T {}

fn _mk() -> Box<dyn Foo> { unimplemented!() }

fn main() {
    *_mk();
}
```

I would be surprised if it could be triggered without `unsized_locals`.

Reply:
I added documentation just for impl trait.  I tried a few times to rewrite the entire section, because the repetition of "expression of an expression statement" was bothering me.  However, each time I tried it seemed to end up worse.  If someone wants to tackle simplifying it, I think that would be good.  Generally it is always an "expression of an expression statement", with different interpretations based on type vs call result.

There are also a lot of other things the `must_use` checks handle, such as uninhabited types and binary operators.  However, that seems to be mostly applicable to std implementations or unstable features.  There is likely more to be said, but I couldn't figure out what should be said.

https://github.com/rust-lang/rust/blob/2c8bbf50db0ef90a33f986ba8fc2e1fe129197ff/src/librustc_lint/unused.rs#L49-L197